### PR TITLE
feat: support misspelled spaced commands

### PR DESCRIPTION
**Pull request by mdonnalley**
Add support for misspelled spaced commands

### CODE REVIEW

File: src/help/util.ts [hunk 1]
```diff
@@ -46,8 +46,8 @@ function collateSpacedCmdIDFromArgs(argv: string[], config: IConfig): string[] {
   const ids = config.commandIDs.concat(config.topics.map(t => t.name))
 
   const findId = (id: string, next: string[]): string | undefined => {
-    const idPresnet = (id: string) => ids.includes(id)
-    if (idPresnet(id) && !idPresnet(`${id}:${next[0]}`)) return id
+    const idPresent = (id: string) => ids.includes(id)
+    if (idPresent(id) && !idPresent(`${id}:${next[0]}`)) return id
```
Comment (RodEsp) [inline src/help/util.ts:50]: Lulz

File: src/main.ts [hunk 2]
```diff
@@ -55,7 +56,24 @@ export async function run(argv = process.argv.slice(2), options?: Interfaces.Loa
   const cmd = config.findCommand(id)
   if (!cmd) {
     const topic = config.findTopic(id)
-    if (topic) return config.runCommand('help', [id])
+    if (topic) {
+      const subCommands = config.commands.filter(c => (c.id).startsWith(id))
+      const argsOfSubCommands = subCommands.reduce((x, y) => x.concat(y.args), [] as Arg[])
```
Comment (RodEsp) [inline src/main.ts:61]: What's the purpose of this reduce? Is it not possible to just cast subCommands to Arg[]?

Not sure of the right TS syntax but something like `const argsOfSubCommands = subCommands as Array<Arg>` or `const argsOfSubCommands = <Array<Arg>>subCommands`?

File: src/main.ts [hunk 3]
```diff
@@ -55,7 +56,24 @@ export async function run(argv = process.argv.slice(2), options?: Interfaces.Loa
...
+
+      if (argsOfSubCommands.length > 0) return config.runCommand('help', [id])
+      if (argvSlice[0].startsWith('--')) return config.runCommand('help', [id])
```
Comment (RodEsp) [inline src/main.ts:64]:
```suggestion
      if (argsOfSubCommands.length > 0 || argvSlice[0].startsWith('--')) return config.runCommand('help', [id])
```

File: src/main.ts (hunk 2 above)
Comment (mdonnalley) [reply src/main.ts:61]: @RodEsp The point is to get a flat list of all args across all the sub commands.

`subCommands` essentially looks like this:

```
[
  { args: [arg1, arg2] },
  {args: [arg3, arg4] }
]
```

And I want this: `[arg1, arg2, arg3, arg4]` which is why I used `reduce`


Regarding the `as Arg[]`: that was required to get the compiler to be okay with accessing the `arg` property within the reducer

File: src/main.ts [hunk 4]
```diff
@@ -55,7 +56,23 @@ export async function run(argv = process.argv.slice(2), options?: Interfaces.Loa
   const cmd = config.findCommand(id)
   if (!cmd) {
     const topic = config.findTopic(id)
-    if (topic) return config.runCommand('help', [id])
+    if (topic) {
+      const subCommands = config.commands.filter(c => (c.id).startsWith(id))
+      const argsOfSubCommands = subCommands.reduce((x, y) => x.concat(y.args), [] as Arg[])
+
+      if (argsOfSubCommands.length > 0 || argvSlice[0].startsWith('--')) return config.runCommand('help', [id])
```
Comment (amphro) [inline src/main.ts:63]: Why does the first arg starting with `--` justify help?

File: src/main.ts (hunk 2 above)
Comment (amphro) [reply src/main.ts:61]: Ya, that `as Arg[]` is very standard. 

The reduce makes sense to flatten, but you are just checking the length? Is the question, "is there one matching command with arguments?" Can't you just say `!!subCommands.find(cmd => cmd.args.length > 0)`

File: src/main.ts [hunk 5]
```diff
@@ -55,7 +56,23 @@ export async function run(argv = process.argv.slice(2), options?: Interfaces.Loa
...
+
+      const extraBits = []
+      for (const arg of argvSlice) {
+        if (arg.startsWith('--')) {
```
Comment (amphro) [inline src/main.ts:67]: Do short chars using `-` matter?

File: src/main.ts [hunk 6]
```diff
@@ -55,7 +56,23 @@ export async function run(argv = process.argv.slice(2), options?: Interfaces.Loa
...
+          break
+        } else {
+          extraBits.push(arg)
+        }
+      }
+      id = [id, ...extraBits].join(':')
```
Comment (amphro) [inline src/main.ts:73]: So basically you are trying to find arguments before the first defined flag and create a new command id. Shouldn't this be what `findCommand` does?

This is hard for me to follow. Can you add some more info in some comments, tests, or the PR description? Maybe we should meet to discuss.

**Review by amphro — CHANGES_REQUESTED** [review]

File: src/help/util.ts [hunk 7]
```diff
@@ -45,14 +45,29 @@ function collateSpacedCmdIDFromArgs(argv: string[], config: IConfig): string[] {
 
   const ids = config.commandIDs.concat(config.topics.map(t => t.name))
 
-  const findId = (id: string, next: string[]): string | undefined => {
-    const idPresnet = (id: string) => ids.includes(id)
-    if (idPresnet(id) && !idPresnet(`${id}:${next[0]}`)) return id
-    if (next.length === 0 || next[0] === '--') return
-    return findId(`${id}:${next[0]}`, next.slice(1))
+  const findId = (argv: string[]): string | undefined => {
```
Comment (amphro) [inline src/help/util.ts:48]: Nice, I think this is really clean.

**Review by amphro — APPROVED** [review]

**Pull request merged**